draco/analysis/hyforesbandpass.py

```diff
@@ -0,0 +1,520 @@
+                    tvar = tools.invert_no_zero(weight[pp, :, xx, tt]) ### inverse of the weight is variance (keep zeros zeros). Looks like the variance is taken over DEC or el.
+
+                    # Grab the filter for this pol and ew baseline
+                    NF = np.ascontiguousarray(filt[pp, :, :, xx, tt])
+
+                    # Make sure that any frequencies unmasked during filter generation
+                    # are also unmasked in the data
+                    valid_freq_flag = np.any(np.abs(NF) > 0.0, axis=0)
+
+                    if not np.any(valid_freq_flag):   ### if the filter is entirely zero --> skip (cond 2/3)
+                        # Skip samples where filter is entirely zero 
+                        weight[pp, :, xx, tt] = 0.0
+                        continue
+
+                    missing_freq = np.flatnonzero(valid_freq_flag & ~flag) ### find the indices of frequencies needed by the delay filter but missing from the unfiltered data --> skip (cond 3/3)
+                    if missing_freq.size > 0:
+                        self.log.warning(
+                            "Missing the following frequencies that were "
+                            "assumed valid during filter generation: "
+                            f"{missing_freq}"
+                        )
+                        weight[pp, :, xx, tt] = 0.0
+                        continue
+
+                    # Apply the filter
+                    post_vis[pp, :, xx, :, tt] = np.matmul(NF, tvis) ### compute using contiguous filter and visibility data ### this must have changed hv.vis
+                    # Do not care about the weight for now
+                    #weight[pp, :, xx, tt] = tools.invert_no_zero( ### variance multiples with the filter value squared
+                        #np.matmul(np.abs(NF) ** 2, tvar)
+                    #)
+                    # Flag frequencies with large attenuation ### mask frequencies where in the filtered data they do not have enough same-frequency contributions from the unfiltered data
+                    if self.atten_threshold > 0.0:
+                        diag = np.abs(np.diag(NF))
+                        nonzero_diag_flag = diag > 0.0
+                        if np.any(nonzero_diag_flag):
+                            med_diag = np.median(diag[nonzero_diag_flag])
+                            flag_low = diag > (self.atten_threshold * med_diag)
+                            weight[pp, :, xx, tt] *= flag_low.astype(weight.dtype) ### this masking is only done on the weight
+                            # Now apply this masking to the filtered visibilities as well
+                            atten_mask = flag_low.astype(vis.dtype)
+                            post_vis[pp, :, xx, :, tt] *= atten_mask[:, np.newaxis]
+
```

Comment:
Rather than have three separate loops over pol, ew baseline, and freq/time, I'm wondering if we could just move everything into the first loop?  I think this would reduce run time, reduce memory usage since you would not have to create a `post_vis` array that is the same shape as the input visibilities, and simplify the code.

I think it would involve inserting something like on this line after you've dealt with the attenuation threshold:
```
post_vis = np.matmul(NF, tvis)
w = (weight[pp, :, xx, tt, np.newaxis] > 0) & el_mask
wtvis = w * tvis
yN[pp, xx, :] += np.sum(wtvis * post_vis, axis=-1)
D[pp, xx, :] += np.sum(wtvis * tvis, axis=-1)
N[pp, xx, :, :] += np.matmul(np.conj(wtvis), wtvis.T) * NF
```